File: Java/main/chap14/javatrampoline/Computation.java
/*
package chap14.javatrampoline;

import java.util.function.Supplier;

public interface Computation<A> {
  A result();

  static <T> Computation<T> done(T value) {
    return new Done<>(value);
  }

  static <T> Computation<T> call(Supplier<Computation<T>> thunk) {
    return new Call<>(thunk);
  }
}

record Done<A>(A value) implements Computation<A> {
  public A result() {
    return value;
  }
}

record Call<A>(Supplier<Computation<A>> thunk) implements Computation<A> {
  public A result() {
    Computation<A> calc = this;
    while (calc instanceof Call<A> call)
      calc = call.thunk().get();
    return calc.result();
  }
}
*/

File: Scala/main/chap14/simpletrampoline/Computation.scala
package chap14.simpletrampoline

import scala.annotation.tailrec
import scala.language.implicitConversions

sealed trait Computation[A]:
   @tailrec
   final def result: A = this match
      case Done(value) => value
      case Call(thunk) => thunk().result

final case class Done[A](value: A)                    extends Computation[A]
final case class Call[A](thunk: () => Computation[A]) extends Computation[A]

implicit def done[A](value: A): Computation[A]       = Done(value)
def call[A](comp: => Computation[A]): Computation[A] = Call(() => comp)
